### PR TITLE
Implementation of delegated routing based on the Edelweiss compiler 

### PR DESCRIPTION
**Pull request by petar**
Ref.
- https://github.com/ipld/edelweiss
- https://github.com/ipld/edelweiss/blob/main/doc/roadmap.md

Resolves https://github.com/ipld/edelweiss/issues/12

### CODE REVIEW

**Comment by BigLep** [comment]
@petar : probably my naivety not to know or see it, but can you provide:
1. command/process you ran to generate this code?  
2. the schema that was used for the generation?

**Comment by petar** [comment]
It’s all in the repo in package gen.

File: gen/routing.go
Comment (willscott) [inline gen/routing.go:131]: Are there docs for this definition format? it seems fairly different from ipld schemas, which is what I was expecting to find here.
I think this is where I'm supposed to comment that there needs to be a graphsync case in addition to bitswap.

More generally, if providers push data to the indexer that is not 'bitswap' (or now '0x900'), how do we represent other transferprotos that have not been explicitly defined in a schema here?

Comment (petar) [reply gen/routing.go:131]: We just finished the RPC compiler, so indeed there is no doc. I am working on it now. "Inductive" is the new name for what used to be called "Union" in IPLD schema. ("Inductive" is the widely-accepted name for this type, "union" is something very different.) Otherwise, the type is the same: You have string-valued tags (the case name), which can carry values of any type.

The transfer protocol differentiator can be implemented in two different ways. Using the "inductive" type:
```go
defs.Inductive{
   Cases: defs.Cases{
      defs.Case{Name: "Bitswap", Type: defs.Ref{Name: "BitswapTransfer"}},
      defs.Case{Name: "Graphsync", Type: defs.Ref{Name: "GraphsyncTransfer"}},
   },
},
```

Or using the "Union" type and the "Singleton" type. The "union" type is a new type (that does not exist in IPLD schema). It differentiates cases based on the (deep) structure of the case subtypes (not based on tags, tag names are just for visual convenience when using unions, they are not reflected on the wire). "Singleton" means that only one specific value can inhabit this type.

```
defs.Union{
      defs.Case{Name: "Bitswap", Type: defs.Ref{Name: "BitswapTransfer"}},
      defs.Case{Name: "Graphsync", Type: defs.Ref{Name: "GraphsyncTransfer"}},
}

Named{
  Name: "BitswapTransfer",
  Type: Structure{
     Fields: Fields{
        Field{Name: "Magic", Type: SingletonInt{0x900}}
        ...
     }
  }
}

Named{
  Name: "GraphsyncTransfer",
  Type: Structure{
     Fields: Fields{
        Field{Name: "Magic", Type: SingletonInt{0xA00}}
        ...
     }
  }
}
```

In this example, the bitswap and the graphsync cases are differentiated because their "magic" fields are always different.

"Union" types are not implemented because they were not part of MVP, however if you end up needing them, they are a couple of days worth of work.

**Comment by BigLep** [comment]
A couple of thoughts here: I'm sure there's legit and genuine curiosity here given that this is some of the first generated code we're seeing come out of Edelweiss (great).  I know @petar needs something to target towards and is using the existing proposed schema.  We ultimately though need to get the schema fully decided as part of https://github.com/ipfs/go-delegated-routing/issues/8.

Thanks @petar for doing some explanation.  I know it's on the Edelweiss project roadmap to be working more on docs and communication this month.  It looks like those will be welcome and appreciated!  Good stuff!

**Comment by BigLep** [comment]
Moving back to in progress because this can't be in review until the spec https://github.com/ipfs/specs/pull/272 is completed.

**Comment by BigLep** [comment]
@petar : thanks for owning this and make the updates for https://github.com/ipfs/specs/pull/272.  I'm assuming once the spec changes fully land, you'll keep updating this.  

Also, given we don't want to spend people's time reviewing the generated code itself, it would be great if the unit test coverage on go-delegated-server was pretty high for client/server communication so that we can have confidence in the generated code as a result.  (I'm saying that as a general statement.  I don't know how high unit test coverage is right now, but if it should have more, let's please include it.)

**Comment by petar** [comment]
@aschmahmann First draft of delegated routing using the new Reframe API is ready. You can do a pass on this PR and then let's regroup to discuss next steps. At the moment, this PR implements the content routing interface which we already use to connect delegated routing into Hydra. As such, this PR is a drop-in replacement for the current delegated routing in master. 

In particular, Get/PutIPNS are not implemented. Also, FindProviders extracts the AddrInfo of all returned provider nodes, regardless of which transport protocols they offer. Perhaps we need to filter out only the bitswap providers, or otherwise we need to extend the Go content routing interface to capture supported transfer protocols.

If I recall correctly, you also want to go a step further and implement a more general content routing interface so that it can be used both in Hydra and IPFS? Let's discuss what that would be.

**Comment by petar** [comment]
@aschmahmann I finished all remaining items we mentioned at standup today:
- FindProviders filters Bitswap nodes
- GetIPNS and PutIPNS included in the Go interface
- Test covers all code-generated code (all method calls)

**Comment by petar** [comment]
Yep, I’ve already filed a PR with unified CI :)

**Comment by petar** [comment]
@aschmahmann I've added unified CI. Improved the regression tests to do multiple tests of different sizes on the same machine and compare the different runs for resource deviations.

File: client/findproviders.go
Comment (aschmahmann) [inline client/findproviders.go:19]: Let's add another wrapper package with some code getting this to implement the relevant libp2p-core routing interfaces (e.g. https://github.com/libp2p/go-libp2p-core/blob/374bb62823579bd73a5e2c6cfaf19de24f6cbd5a/routing/routing.go#L26 and https://github.com/libp2p/go-libp2p-core/blob/374bb62823579bd73a5e2c6cfaf19de24f6cbd5a/routing/routing.go#L49).

This will make it easier for people to use tools like https://github.com/libp2p/go-libp2p-routing-helpers/blob/506670d53a31503f257980dbf354b3ea1b990fb7/parallel.go#L20

**Review by aschmahmann — COMMENTED** [review]
Overall looks good. Left some comments/questions as well as feedback on the pieces we're going to want to get this usable within go-ipfs.

Lmk if the parts where I've requested more testing are a pain

File: server/findproviders.go
Comment (petar) [reply server/findproviders.go:22]: Both conventions would work for the typical use case. However, this convention provides strictly more flexibility to the caller of the interface. E.g. it allows the caller to reuse existing channels instead of creating new ones in tricker applications.

File: client/findproviders.go
Comment (petar) [reply client/findproviders.go:13]: Done. I am excluding the `-` in order for this logger to be identical to the logger generated by the compiler. (The compiler generates logger names as strings.ToLower(..) of whatever the service name is in the definition. The latter is also used as in the Go naming of the client and service types.)

File: server/findproviders.go
Comment (petar) [reply server/findproviders.go:18]: Identify is handled in the auto-generated code. Regarding client behavior:
I don't think that we should bake in an initial call to Identify by default. This would add a roundtrip to use cases where a client is used only once. Come to think of it, if we want to "cache" what methods are available, we should do it lazily: make an optimistic call to some method and then remember if it worked out or not (for future calls). That latter functionality does not belong here anyway. It is generic, so it should live in Edelweiss. What do you think?

File: client/findproviders.go
Comment (petar) [reply client/findproviders.go:26]: I've added a wrapper for the ContentRouting interface.

Are you sure you want a wrapper for the ValueStore interface? This interface doesn't seem to map cleanly to the FindProviders/GetIPNS/PutIPNS interface. The Get path would require trying to decode the stored record either as a provider or as an IPNS record. Is there a use case for this?

Also, let's not extend the scope of this PR beyond what is necessary for Hydra and the Delegated Routing code path in IPFS.

File: test/clientserver_test.go
Comment (petar) [reply test/clientserver_test.go:154]: > We're pretty light on unit/correctness tests here and ofc the 4k LoC generated code isn't really human-examinable. Maybe the appropriate tests live in Edelweiss but having a few more here would be great.
> 
> Some of the edge cases that seem worth checking if not too painful include:
> 
> * How things work if I the client sends the server data not covered by the schema (i.e. does it fail, or does the base message make it through?). What about when the server responds with data not covered by the server?
This is in the spec, no? If you send more data than you need (either for client or server), things proceed successfully.
If data from the schema is missing, an error is returned. These tests belong in Edelweiss, since they are not specific to this API.

> * What about new/missing functions?
You get an error if you call a missing function. Since this is also generic, test belongs in Edelweiss. Or do you mean you want to check how the errors surface through the client/server wrappers?

> * What happens if a returned multiaddr ends in `/p2p/peerID` does the client error?
Not currently. What is the desired behavior?

> * Are our `Any` fallbacks working properly?
You mean, skipping unknown transfer protocols and unknown node types? I'll add tests.

> * Are we passing `Error` responses correctly?
I am guessing you mean if a server-side returns asynchronous error, that they show up as such on the client side. This would be tested in Edelweiss at the generated code level of client and server. If we want it all the way through to the wrapper client and server, I'll add tests in this repo.

It sounds like adding tests on the wrapper client/server will make everyone feel good. With a lot of these tests there are two levels of possible testing: generate client to generated server, or wrapped client to wrapped server. The former should be in Edelweiss and the latter here (obviously).

File: go.mod
Comment (aschmahmann) [reply go.mod:9]: I meant the go-ipld-prime version not edelweiss

File: server/findproviders.go
Comment (aschmahmann) [reply server/findproviders.go:22]: > E.g. it allows the caller to reuse existing channels instead of creating new ones in trickier applications.

Does it? If the interface declares that the channel is closed by the functions (e.g. `FindProviders`) then the channel can't be reused. If the interface declares that the channel is closed by the caller then there's no way for the caller to know when it should close the channel.

What were you envisioning here?

File: server/findproviders.go
Comment (aschmahmann) [reply server/findproviders.go:18]: Do we have a way of knowing that a given method is unsupported rather than there being some intermittent issue?

Also, probably obvious but in the identify case we could save the round-trip by querying in parallel. It would add an extra query (particularly annoying in the repeated single-use case), but without the added latency.

File: test/clientserver_test.go
Comment (aschmahmann) [reply test/clientserver_test.go:154]: > What happens if a returned multiaddr ends in /p2p/peerID does the client error?
>> Not currently. What is the desired behavior?

[The spec](https://github.com/ipfs/specs/blob/6bdb7b2751038e1d0212a40494ea8fd4018f384c/REFRAME.md?plain=1#L168) indicates we shouldn't end in `/p2p/peerID` so our options seem to be:
1. Be very strict and return an error if any part of the response is non-compliant
2. Be a little strict and drop multiaddrs with `/p2p/peerID` at the end perhaps with some log output informing about the error
3. Be loose and rewrite the multiaddr to not end with `/p2p/peerID` assuming the extra peerID matches the peerID in the `ID` field. If it doesn't match then log.

2 seems reasonable to me WDYT?

File: server/findproviders.go
Comment (petar) [reply server/findproviders.go:18]: The error indicating a method is missing or the schema is unsupported is returned in the first asynchronous result. I can:
- make these errors distinguishable (from others)
- update the generated client code to cache the information that a method is missing (and not attempt further calls and return an error to the user right away)

This solution would remove the need for calling Identify. @aschmahmann  What do you think?

Comment (aschmahmann) [reply server/findproviders.go:18]: Seems like a reasonable strategy, although I wonder if `Identify` is still a helpful thing to have around whether for debugging or as an alternative mechanism for figuring out what's supported.

If it seems unhelpful/wasteful we could go back to the spec and drop it given this is the first implementation. WDYT?

Comment (petar) [reply server/findproviders.go:18]: I would keep it for now. I see a potential/eventual/useful extension of the Identify method's response which returns the schema (request and response) for each supported method.

File: client/findproviders.go
Comment (petar) [reply client/findproviders.go:26]: Makes sense. Let me open a new issue about adding the `ValueStore` interface. I want to unblock the integration with hydra for now, without expanding scope since it's been too long this issue has been open. Added https://github.com/ipfs/go-delegated-routing/issues/13

File: test/clientserver_test.go
Comment (petar) [reply test/clientserver_test.go:154]: Yes, I think 2 is the way to go.

File: server/findproviders.go
Comment (petar) [inline server/findproviders.go:128]: @aschmahmann Please, verify this is as intended. PeerAddrs are created from whatever the server-implementing user passes. The code rewrites the multiaddresses to make sure they conform to the spec.

File: client/findproviders.go
Comment (petar) [inline client/findproviders.go:110]: @aschmahmann Please, verify this is as intended.

**Comment by petar** [comment]
@aschmahmann I've addressed all the comments. Please take another look.

File: client/findproviders.go
Comment (aschmahmann) [inline client/findproviders.go:18]: This should be validated if it's going to substitute into the `ValueStore` interface and run alongside the DHT and PubSub in go-ipfs.

See https://github.com/ipfs/go-namesys/blob/bf4b3cffd9bdc0fb668bb6879f17070128145798/routing.go#L82 which indicates that the underlying system is responsible for both validating and only returning increasingly recent records.

DHT Get Validation: https://github.com/libp2p/go-libp2p-kad-dht/blob/0b7ac010657443bc0675b3bd61133fe04d61d25b/routing.go#L320

DHT Get Ordering (so that in the async case only increasingly good records are returned): https://github.com/libp2p/go-libp2p-kad-dht/blob/0b7ac010657443bc0675b3bd61133fe04d61d25b/routing.go#L228.

---

If we weren't going to validate here then we'd need really big warning lights like calling the function `GetIPNSUnvalidated` or something to prevent mistakes, however it's easy enough to add the validators from go-ipns so why not 🤷.

File: client/contentrouting.go
Comment (aschmahmann) [inline client/contentrouting.go:30]: Thoughts (not blockers for merge):

Not sure if this is avoidable but wanted your thoughts on this. It looks like every abstraction layer involving a channel turns into another goroutine to pull `x` from the channel and return `f(x)` into a new channel.

Maybe once generics become more widely supported in our codebase (once Go 1.19 is released) this will become easier by letting us return generic channels and pass in functions to convert between types so a goroutine isn't required.

To be fair short lived goroutines aren't necessarily a big issue unless we're doing them on mass so we can cross this bridge when it becomes a problem.

**Review by aschmahmann — CHANGES_REQUESTED** [review]
Looks good! The main comments are around lack of validation of the responses we get as a client and missing chunks of the IPNS records.

For the IPNS records I'm fine if you'd prefer to split it into a different PR, but I don't think it'll be too much if you want to do it here (i.e. do validation and some basic testing).

File: client/contentrouting.go
Comment (aschmahmann) [inline client/contentrouting.go:11]: Go certainly doesn't require this but if you want to assert interface compliance with one or more of the routing interfaces from https://github.com/libp2p/go-libp2p-core/blob/525a0b13017263bde889a3295fa2e4212d7af8c5/routing/routing.go#L67 then you can do so with the pattern `var _ MyInterface = (*myStruct)(nil)`.

This means you won't need accidentally break an interface without realizing it.

File: client/contentrouting_test.go
Comment (petar) [reply client/contentrouting_test.go:49]: I've added a comment to explain this. This test is dedicated to checking that ContentRoutingClient.FindProvidersAsync correctly wraps DelegatedRoutingClient.FindProvidersAsync, purely in terms of the async semantics of returned results (values are ignored). The two methods have different semantics (one has maxValues, the other doesn't). This test focuses on making sure that the correct number of results are returned.

The actual implementation of DelegatedRoutingClient interface in terms of an Edelweiss network service has checks for correct values. These are in the clientserver_test.go.

File: client/contentrouting_test.go
Comment (petar) [reply client/contentrouting_test.go:38]: See my response to the previous comment. Get/PutIPNS are not present in ContentRouting interface, so there is nothing to test here.

File: client/findproviders.go
Comment (petar) [reply client/findproviders.go:18]: @aschmahmann Should the ipns Validator be passed as an option when a delegated routing client is created?

Comment (aschmahmann) [reply client/findproviders.go:18]: It could, although realistically nobody passes in `KeyBooks` anymore (https://github.com/ipfs/go-ipns/blob/33ef6df0c42c6f1c736025dfa2ceecbd064e9dff/record.go#L31) so you could just construct an `ipns.Validator{}` here without any options.

File: test/clientserver_test.go
Comment (petar) [inline test/clientserver_test.go:219]: @aschmahmann This code here doesn't seem to generate a correct pair of IPNS key and record. Can you advise what is wrong?

File: go.mod
Comment (aschmahmann) [inline go.mod:7]: latest tag is v0.1.2

File: go.mod
Comment (aschmahmann) [inline go.mod:12]: latest is v0.1.3

File: go.mod
Comment (aschmahmann) [inline go.mod:13]: latest is v0.5.0

File: test/clientserver_test.go
Comment (aschmahmann) [reply test/clientserver_test.go:219]: The record is invalid because it's expired (you passed an EOL of `time.Now()`, note that the 1hr you passed was the TTL/how long the author thinks it's safe to cache records for as with DNS).

However, some issues in the code in the implementation here made surfacing this error to you more difficult (i.e. the only way you would've noticed the error was by reading an Info log). After these are fixed we should probably add a test for the expired/invalid records just to make sure we've bubbled the errors up correctly (i.e. to cover any changes we're making)

File: client/getipns.go
Comment (aschmahmann) [inline client/getipns.go:21]: Calling `Select` on an empty set errors
https://github.com/ipfs/go-ipns/blob/5976a80227cc5199414119585cca347bc814647a/record.go#L106.

We should probably return the standard error for a missing record when the set is empty https://github.com/libp2p/go-libp2p-core/blob/02cbdcc4191b0f512c6cf4f2b1a8eeaaf32f1061/routing/routing.go#L15

File: client/getipns.go
Comment (aschmahmann) [inline client/getipns.go:20]: Can this ever happen? The validator is built in now so seems like we can remove this condition.

File: client/getipns.go
Comment (aschmahmann) [inline client/getipns.go:27]: If we're removing this if statement then we should probably change GetIPNS to return a single record `[]byte` instead of a set `[][]byte` since we don't have anything interesting to do with the worse/older records.

File: client/getipns.go
Comment (aschmahmann) [inline client/getipns.go:56]: same question as above

File: client/getipns.go
Comment (aschmahmann) [inline client/getipns.go:58]: Why not return an `r1.Err` down the channel here instead of just logging?

**Review by aschmahmann — CHANGES_REQUESTED** [review]

File: client/getipns.go
Comment (petar) [reply client/getipns.go:58]: That's fine too. I was debating whether to mix errors produced on the client-side with errors coming from the server side. I've updated it.

File: client/contentrouting.go
Comment (petar) [reply client/contentrouting.go:32]: Yes. Without generics, we would need to resort to empty interfaces, which I think makes the code very unmanageable. I lean towards code readability unless there is a clear performance problem. And indeed it seems like Go generics can fix this issue.

**Comment by petar** [comment]
@aschmahmann It's worth another look. I think I've addressed all comments in this round.

**Comment by petar** [comment]
@aschmahmann Here's what I am thinking should be next steps:
- I'll make a release on Edelweiss at its current commit
- Bump the dependencies here on the Edelweiss release
- Merge this PR
Then:
- Start a Hydra PR that utilizes this PR
- Then we need to discuss re-deployment of Hydra
Sound good?

**Comment by aschmahmann** [comment]
@petar sounds like a plan to me 😄. In order for the hydras to properly utilize it there needs to be some server responding to Reframe requests.

However, you might save yourself some pain by working on getting the indexers to serve the data using Reframe earlier. Otherwise you'll have to get the Hydras to support both the Indexer and Reframe endpoints.

Your call though, if you think it'll be easier to start off working in the Hydras then start there.

**Comment by petar** [comment]
> @petar sounds like a plan to me 😄. In order for the hydras to properly utilize it there needs to be some server responding to Reframe requests.
> 
> However, you might save yourself some pain by working on getting the indexers to serve the data using Reframe earlier. Otherwise you'll have to get the Hydras to support both the Indexer and Reframe endpoints.
> 
> Your call though, if you think it'll be easier to start off working in the Hydras then start there.

Sounds good! Thanks for reviewing this monster PR :)

**Pull request merged**
